helix-core/src/main/java/org/apache/helix/messaging/handling/HelixStateTransitionHandler.java

```diff
@@ -463,6 +442,36 @@ public void onError(Exception e, ErrorCode code, ErrorType type) {
 
   }
 
+  // Verify the fromState and current state of the stateModel.
+  private void validateStaleMessage (boolean isPreCheck) throws Exception {
+    String fromState = _message.getFromState();
+    String toState = _message.getToState();
+    String partitionName = _message.getPartitionName();
+
+    // state in _currentStateDelta uses current state from state model. It has the
+    // most up-to-date. current state. In case currentState in stateModel is null,
+    // partition is in initial state and we using it as current state.
+    // Defined in HelixStateMachineEngine.
+    String state = _currentStateDelta.getState(partitionName);
+
+    if (toState.equalsIgnoreCase(state)) {
+      // To state equals current state, we can just ignore the message
+      throw new HelixDuplicatedStateTransitionException(String
```

Comment:
This is not duplicate state transition. Could be some STs are old in pending queue just not match what target state we have now.

Reply:
If it already been processed (in READ or UNPROCESSABLE state) then we would ignore the ST message before this check. (line 856 in HelixTaskExecutor)

Reply:
What I am trying to say is not use HelixDuplicatedStateTransitionException, because there could be other cases for this scenario. If we check log, we may get confused.

Reply:
Thanks for the follow up. 
In my understanding, this message must be a new message generated by controller, maybe caused by reading the old currentState etc.Since this is a new message, doesn't this count as HelixDuplicatedStateTransition?